File: aspnet-in-practice/07/VB/AdaptiveRendering/App_Code/DropDownListAdapter.vb
Imports System
Imports System.Web
Imports System.Web.UI.WebControls.Adapters
Imports System.Web.UI
Imports System.Web.UI.WebControls
Imports System.Collections.Generic

Namespace ASPNET4InPractice
    Public Class DropDownListAdapter
        Inherits WebControlAdapter
        Private uniqueID As String = Nothing
        
        Protected Overloads Overrides Sub RenderContents(ByVal writer As HtmlTextWriter)
            Dim list As DropDownList = TryCast(Me.Control, DropDownList)
            
            uniqueID = list.UniqueID
            
            Dim lastOptionGroup As String = Nothing
            Dim currentOptionGroup As String = Nothing
            For Each item As ListItem In list.Items
                currentOptionGroup = TryCast(item.Attributes("OptionGroup"), String)
                
                If currentOptionGroup IsNot Nothing Then
                    If lastOptionGroup Is Nothing OrElse Not lastOptionGroup.Equals(currentOptionGroup, StringComparison.InvariantCultureIgnoreCase) Then
                        If lastOptionGroup IsNot Nothing Then
                            RenderOptionGroupEndTag(writer)
                        End If
                        
                        RenderOptionGroupBeginTag(currentOptionGroup, writer)
                    End If
                    
                    lastOptionGroup = currentOptionGroup
                End If
                
                RenderListItem(item, writer)
            Next
            
            If lastOptionGroup IsNot Nothing Then
                RenderOptionGroupEndTag(writer)
            End If
        End Sub
        
        Private Sub RenderOptionGroupBeginTag(ByVal name As String, ByVal writer As HtmlTextWriter)
            writer.Indent += 1
            writer.WriteBeginTag("optgroup")
            writer.WriteAttribute("label", name)
            writer.Write(HtmlTextWriter.TagRightChar)
            writer.WriteLine()
        End Sub
        
        Private Sub RenderOptionGroupEndTag(ByVal writer As HtmlTextWriter)
            writer.WriteEndTag("optgroup")
            writer.WriteLine()
            writer.Indent -= 1
        End Sub
        
        Private Sub RenderListItem(ByVal item As ListItem, ByVal writer As HtmlTextWriter)
            writer.Indent += 1
            writer.WriteBeginTag("option")
            writer.WriteAttribute("value", item.Value, True)
            
            If item.Selected Then
                writer.WriteAttribute("selected", "selected", False)
            End If
            
            For Each key As String In item.Attributes.Keys
                If Not key.Equals("optiongroup", StringComparison.CurrentCultureIgnoreCase) Then
                    writer.WriteAttribute(key, item.Attributes(key))
                End If
            Next
            
            writer.Write(HtmlTextWriter.TagRightChar)
            
            If Page IsNot Nothing Then
                Page.ClientScript.RegisterForEventValidation(uniqueID, item.Value)
            End If
            
            HttpUtility.HtmlEncode(item.Text, writer)
            writer.WriteEndTag("option")
            writer.WriteLine()
            writer.Indent -= 1
        End Sub
    End Class
End Namespace

File: aspnet-in-practice/07/CS/AdaptiveRendering/App_Code/DropDownListAdapter.cs
using System;
using System.Web;
using System.Web.UI.WebControls.Adapters;
using System.Web.UI;
using System.Web.UI.WebControls;
using System.Collections.Generic;

namespace ASPNET4InPractice
{
    public class DropDownListAdapter : WebControlAdapter
    {
        string uniqueID = null;

        protected override void RenderContents(HtmlTextWriter writer)
        {
            DropDownList list = this.Control as DropDownList;

            uniqueID = list.UniqueID;

            string lastOptionGroup = null;
            string currentOptionGroup = null;
            foreach (ListItem item in list.Items)
            {
                currentOptionGroup = item.Attributes["OptionGroup"] as string;

                if (currentOptionGroup != null)
                {
                    if (lastOptionGroup == null || !lastOptionGroup.Equals(currentOptionGroup, StringComparison.InvariantCultureIgnoreCase))
                    {
                        if (lastOptionGroup != null)
                            RenderOptionGroupEndTag(writer);

                        RenderOptionGroupBeginTag(currentOptionGroup, writer);
                    }

                    lastOptionGroup = currentOptionGroup;
                }

                RenderListItem(item, writer);
            }

            if (lastOptionGroup != null)
                RenderOptionGroupEndTag(writer);
        }

        private void RenderOptionGroupBeginTag(string name, HtmlTextWriter writer)
        {
            writer.Indent++;
            writer.WriteBeginTag("optgroup");
            writer.WriteAttribute("label", name);
            writer.Write(HtmlTextWriter.TagRightChar);
            writer.WriteLine();
        }

        private void RenderOptionGroupEndTag(HtmlTextWriter writer)
        {
            writer.WriteEndTag("optgroup");
            writer.WriteLine();
            writer.Indent--;
        }

        private void RenderListItem(ListItem item, HtmlTextWriter writer)
        {
            writer.Indent++;
            writer.WriteBeginTag("option");
            writer.WriteAttribute("value", item.Value, true);

            if (item.Selected)
                writer.WriteAttribute("selected", "selected", false);

            foreach (string key in item.Attributes.Keys)
            {
                if (!key.Equals("optiongroup", StringComparison.CurrentCultureIgnoreCase))
                    writer.WriteAttribute(key, item.Attributes[key]);
            }

            writer.Write(HtmlTextWriter.TagRightChar);

            if (Page != null)
            {
                Page.ClientScript.RegisterForEventValidation(uniqueID, item.Value);
            }

            HttpUtility.HtmlEncode(item.Text, writer);
            writer.WriteEndTag("option");
            writer.WriteLine();
            writer.Indent--;
        }
    }
}